### PR TITLE
Fix some clang-tidy performance warnings

### PR DESCRIPTION
**Pull request by DanielChabrowski**
Hello,
I run some clang-tidy performance-* checks on the code and decided to correct the warnings. Most of them warn about unnecessary copy or to preallocate a vector.

It seems that clang-format-8 added some additional unintended changes.

Cheers.

### CODE REVIEW

File: src/clang_tu.cc [hunk 1]
```diff
@@ -104,7 +104,7 @@ BuildCompilerInvocation(const std::string &main, std::vector<const char *> args,
       CompilerInstance::createDiagnostics(new DiagnosticOptions,
                                           new IgnoringDiagConsumer, true));
   std::unique_ptr<CompilerInvocation> CI =
-      createInvocationFromCommandLine(args, Diags, VFS);
+      createInvocationFromCommandLine(args, Diags, std::move(VFS));
```
Comment (MaskRay) [inline src/clang_tu.cc]: This is not very necessary because VFS itself is just a ref counter..

File: src/lsp.cc [hunk 2]
```diff
@@ -63,7 +63,7 @@ void DocumentUri::SetPath(const std::string &path) {
   // file:///c%3A/Users/jacob/Desktop/superindex/indexer/full_tests
   raw_uri = path;
 
-  size_t index = raw_uri.find(":");
+  size_t index = raw_uri.find(':');
```
Comment (MaskRay) [inline src/lsp.cc]: This is fine

File: src/message_handler.cc [hunk 3]
```diff
@@ -228,7 +227,8 @@ void MessageHandler::Run(InMessage &msg) {
       } catch (NotIndexed &) {
         throw;
       } catch (...) {
-        reply.Error(ErrorCode::InternalError, "failed to process " + msg.method);
+        reply.Error(ErrorCode::InternalError,
```
Comment (MaskRay) [inline src/message_handler.cc]: These intermixed formatting is not very necessary..

File: src/message_handler.cc [hunk 4]
```diff
@@ -121,9 +121,8 @@ void ReplyOnce::ReplyLocationLink(std::vector<LocationLink> &result) {
   if (g_config->client.linkSupport) {
     (*this)(result);
   } else {
-    std::vector<Location> result1;
-    for (auto &loc : result)
-      result1.emplace_back(std::move(loc));
+    std::vector<Location> result1{std::make_move_iterator(result.begin()),
```
Comment (MaskRay) [inline src/message_handler.cc]: This is good.

File: src/project.hh [hunk 5]
```diff
@@ -76,7 +76,7 @@ struct Project {
   void SetArgsForFile(const std::vector<const char *> &args,
                       const std::string &path);
 
-  void Index(WorkingFiles *wfiles, RequestId id);
+  void Index(WorkingFiles *wfiles, const RequestId &id);
```
Comment (MaskRay) [inline src/project.hh]: This is fallout of https://github.com/MaskRay/ccls/issues/437 ..

File: src/query.cc [hunk 6]
```diff
@@ -357,7 +357,7 @@ void DB::ApplyIndexUpdate(IndexUpdate *u) {
       RefDecl(prev_lid2file_id, usr, Kind::Func, *def.spell, -1);
   RemoveUsrs(Kind::Func, u->file_id, u->funcs_removed);
   Update(lid2file_id, u->file_id, std::move(u->funcs_def_update));
-  for (auto &[usr, del_add]: u->funcs_declarations) {
+  for (auto &[usr, del_add] : u->funcs_declarations) {
```
Comment (MaskRay) [inline src/query.cc]: Unnecessary formatting changes

File: src/query.cc [hunk 7]
```diff
@@ -476,7 +476,7 @@ void DB::Update(const Lid2file_id &lid2file_id, int file_id,
 void DB::Update(const Lid2file_id &lid2file_id, int file_id,
                 std::vector<std::pair<Usr, QueryVar::Def>> &&us) {
   for (auto &u : us) {
-    auto &def = u.second;
+    auto def = u.second;
```
Comment (MaskRay) [inline src/query.cc]: This is wrong.

File: src/sema_manager.cc [hunk 8]
```diff
@@ -245,14 +245,13 @@ class StoreDiags : public DiagnosticConsumer {
   void BeginSourceFile(const LangOptions &Opts, const Preprocessor *) override {
     LangOpts = &Opts;
   }
-  void EndSourceFile() override {
```
Comment (MaskRay) [inline src/sema_manager.cc]: Unnecessary formatting.

File: src/serializer.cc [hunk 9]
```diff
@@ -35,7 +35,7 @@ bool gTestOutputMode = false;
 
 namespace ccls {
 
-void JsonReader::IterArray(std::function<void()> fn) {
+void JsonReader::IterArray(const std::function<void()> &fn) {
```
Comment (MaskRay) [inline src/serializer.cc]: I would change this to use function_ref instead

File: src/test.cc [hunk 10]
```diff
@@ -15,16 +15,16 @@ limitations under the License.
```
Comment (MaskRay) [inline src/test.cc]: test.cc is under maintenance and I'd probably remove it. So further changes to this file is not needed.

File: src/message_handler.cc (hunk 3 above)
Comment (DanielChabrowski) [reply src/message_handler.cc]: The project isn't formatted properly therefore the changes. I can try to remove formatting from this PR and make another one with pure clang-format run on src directory. Would that be okay?

**Comment by MaskRay** [comment]
Since there are many accumulated formatting problems and I'd like to change variable/function cases, I'll do an enormous formatting change: https://github.com/MaskRay/ccls/pull/471

**Comment by DanielChabrowski** [comment]
> Since there are many accumulated formatting problems and I'd like to change variable/function cases, I'll do an enormous formatting change: #471

Did you update .clang-format config? It still modifies pretty much every file for me. Tried both clang-format 7 and 8.

**Comment by MaskRay** [comment]
> > Since there are many accumulated formatting problems and I'd like to change variable/function cases, I'll do an enormous formatting change: #471
> 
> Did you update .clang-format config? It still modifies pretty much every file for me. Tried both clang-format 7 and 8.

I ran `git diff -U0 --no-color 'HEAD^' | ~/llvm/tools/clang/tools/clang-format/clang-format-diff.py -i -p1` to format lines touched by the variable/function renaming change.
I am using clang-format built from trunk. It may be different a bit from clang-format 7 and 8.

**Comment by XVilka** [comment]
There are a lot of conflicts now that should be resolved.

File: src/serializer.hh [hunk 11]
```diff
@@ -48,8 +49,8 @@ struct JsonReader {
   JsonReader(rapidjson::Value *m) : m(m) {}
   void startObject() {}
   void endObject() {}
-  void iterArray(std::function<void()> fn);
-  void member(const char *name, std::function<void()> fn);
+  void iterArray(const llvm::function_ref<void()> fn);
```
Comment (MaskRay) [inline src/serializer.hh]: Remove `const `

File: src/serializer.cc [hunk 12]
```diff
@@ -48,7 +48,7 @@ void JsonReader::iterArray(std::function<void()> fn) {
   }
   path_.pop_back();
 }
-void JsonReader::member(const char *name, std::function<void()> fn) {
+void JsonReader::member(const char *name, const llvm::function_ref<void()> fn) {
```
Comment (MaskRay) [inline src/serializer.cc]: Remove `const `

File: src/serializer.cc [hunk 13]
```diff
@@ -35,7 +35,7 @@ bool gTestOutputMode = false;
 
 namespace ccls {
 
-void JsonReader::iterArray(std::function<void()> fn) {
+void JsonReader::iterArray(const llvm::function_ref<void()> fn) {
```
Comment (MaskRay) [inline src/serializer.cc]: Remove `const `

File: src/sema_manager.cc [hunk 14]
```diff
@@ -231,7 +231,7 @@ class StoreDiags : public DiagnosticConsumer {
   }
 
 public:
-  StoreDiags(std::string path) : path(path) {}
+  StoreDiags(std::string path) : path{std::move(path)} {}
```
Comment (MaskRay) [inline src/sema_manager.cc]: `path(std::move(path))`

File: src/pipeline.cc [hunk 15]
```diff
@@ -717,8 +717,9 @@ void standalone(const std::string &root) {
   WorkingFiles wfiles;
   VFS vfs;
   SemaManager manager(
-      nullptr, nullptr, [&](std::string, std::vector<Diagnostic>) {},
-      [](RequestId id) {});
+      nullptr, nullptr,
+      [&](const std::string &, const std::vector<Diagnostic> &) {},
```
Comment (MaskRay) [inline src/pipeline.cc]: Delete `&`

File: src/message_handler.cc (hunk 4 above)
Comment (MaskRay) [reply src/message_handler.cc]: Use `(..)` instead of `{...}`. `{}` can be confusing because the constructor has an overload on initializer_list.

File: src/pipeline.cc [hunk 16]
```diff
@@ -616,13 +616,13 @@ void mainLoop() {
 
   SemaManager manager(
       &project, &wfiles,
-      [&](std::string path, std::vector<Diagnostic> diagnostics) {
+      [&](const std::string &path, std::vector<Diagnostic> diagnostics) {
```
Comment (MaskRay) [inline src/pipeline.cc]: `[&]` -> `[]`

Comment (MaskRay) [reply src/pipeline.cc]: sema_manager.hh `OnDiagnostic` can be changed to llvm::function_ref now.

Comment (DanielChabrowski) [reply src/pipeline.cc]: Lambda capture done. OnDiagnostic & OnDropped part can be delivered by a different PR.

**Comment by MaskRay** [comment]
Thanks a lot! (and bear with me making it hard to rebase). I'll make some more adjustments and merge.

The below is a false positive.

```cpp
src/utils.cc:155:7: warning: string concatenation results in allocation of unnecessary temporary strings; consider using 'operator+=' or 'string::append()' instead [performance-inefficient-string-concatenation]
      path = root + path.substr(real.size());
```

Changing it to `string::append` will destroy `root`.

`serializer.hh` should not include too many llvm headers. `#include <llvm/ADT/STLExtras.h>` should be moved to `serializer.cc`
